contents/java/Java.md

```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+
+One important point is that while Java reflections are powerful, its implementations are anything but. There are however
+some libraries out there to make life easier such as the [Google's Guava
+library](https://github.com/google/guava/wiki/ReflectionExplained) which contains many utility methods and classes to
+make your life easier.
+
+#### Applications
+
+You might have learnt from your Software Engineering module that the Observer pattern can be used for Objects that are
+interested to get notified if the state of another object is changed. However, a you don't want to create a troublesome
+bidirectional dependency between two unrelated objects that have no business talking to each other. This is when the
+Observer pattern becomes useful.
+
+One prime example of the implementation of the `Observer` pattern is the Google Events bus used in `AddressBook Level 4`.
+The event bus uses reflections to observe all registered objects via `register` method for methods annotated with the
+`Subscribe` annotation.
+
+An example implementation (not the actual) of the `Subscribe` annotation might be:
+
+```java
+@Retention(RetentionPolicy.RUNTIME) // Retain this annotation at runtime!
+@Target(ElementType.METHOD) // Only can be applied to methods!
+public @interface Subscribe { }
+```
+
+And that is it! Important parts of the code to note is the first and second line before the declaration. The first line
+tells Java that this annotation must not be discarded during compile time so it will be available during runtime. The
+retention poilcy is there because some annotation do not mean anything after compilation (such as `Override` and
+`SuppressWarnings`), so it does not make sense to keep it after compiling.
+
+The second line just means that this annotation can be applied to methods only.
+
+And the more important part is the Subscriber Registry
+
+```java
+// TypeToken class is provided by Google Guava reflection library
+Set<? extends Class<?>> supertypes = TypeToken.of(clazz).getTypes().rawTypes();
+
+for (Class<?> supertype : supertypes) {
+      for (Method method : supertype.getDeclaredMethods()) {
+        if (method.isAnnotationPresent(Subscribe.class) {
+            registerSubscriber(method);
+        }
+    }
+}
+```
+
+The first line finds all classes and its parent classes of the registered class and converts it to a set. That is if you
+registered `Sheep extends Animal` as an event handler, both `Sheep` and `Animal` will be captured by the first line. The
+following lines will then examine all their methods (during runtime!) for the subscribe annotation and register it so
+that it will receive the specified event when it is fired. Of course this implementation leaves out a lot of details but
+you get the idea of how java reflections
+
+### Disadvantages of reflections
+
+While Java reflection is powerful, you should not immediately jump on the reflections ship. This is because there are
```

Comment:
Maybe write "Java reflections" instead of "Java reflection" to be consistent with the plural form used throughout the chapter.